ratelimit/redis.go

```diff
@@ -0,0 +1,247 @@
+package ratelimit
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/go-redis/redis"
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/metrics"
+)
+
+// RedisOptions is used to configure the redis.Ring
+type RedisOptions struct {
+	// Addrs are the list of redis shards
+	Addrs []string
+	// ReadTimeout for redis socket reads
+	ReadTimeout time.Duration
+	// WriteTimeout for redis socket writes
+	WriteTimeout time.Duration
+	// PoolTimeout is the max time.Duration to get a connection from pool
+	PoolTimeout time.Duration
+	// MinIdleConns is the minimum number of socket connections to redis
+	MinIdleConns int
+	// MaxIdleConns is the maximum number of socket connections to redis
+	MaxIdleConns int
+}
+
+type ring struct {
+	ring    *redis.Ring
+	metrics metrics.Metrics
+}
+
+// clusterLimitRedis stores all data required for the cluster ratelimit.
+type clusterLimitRedis struct {
+	group   string
+	maxHits int64
+	window  time.Duration
+	ring    *redis.Ring
+	metrics metrics.Metrics
+}
+
+const (
+	DefaultReadTimeout  = 25 * time.Millisecond
+	DefaultWriteTimeout = 25 * time.Millisecond
+	DefaultPoolTimeout  = 25 * time.Millisecond
+	DefaultMinConns     = 100
+	DefaultMaxConns     = 100
+
+	defaultConnMetricsInterval = 60 * time.Second
+	redisMetricsPrefix         = "swarm.redis."
+)
+
+func newRing(ro *RedisOptions, quit <-chan struct{}) *ring {
+	var r *ring
+
+	// ring.Options())
+	// HeartbeatFrequency:500000000,MaxRetries:0, MinRetryBackoff:8000000, MaxRetryBackoff:512000000, DialTimeout:0, ReadTimeout:0, WriteTimeout:0, PoolSize:0, MinIdleConns:0, MaxConnAge:0, PoolTimeout:0, IdleTimeout:0, IdleCheckFrequency:0}"
+	ringOptions := &redis.RingOptions{
+		Addrs: map[string]string{},
+	}
+
+	if ro != nil {
```

Comment:
it doesn't make sense to call `newRing` when `ro == nil`, and it's a package-internal function